lib/rdkafka/consumer.rb

```diff
@@ -441,5 +441,71 @@ def each
         end
       end
     end
+
+    # Poll for new messages and yield them in batches that may contain
+    # messages from more than one partition.
+    #
+    # Rather than yield each message immediately as soon as it is received,
+    # each_batch will attempt to wait for as long as `timeout_ms` in order
+    # to create a batch of up to but no more than `max_items` in size.
+    #
+    # Said differently, if more than `max_items` are available within
+    # `timeout_ms`, then `each_batch` will yield early with `max_items` in the
+    # array, but if `timeout_ms` passes by with fewer messages arriving, it
+    # will yield an array of fewer messages, quite possibly zero.
+    #
+    # In order to prevent wrongly auto committing many messages at once across
+    # possibly many partitions, callers must explicitly indicate which messages
+    # have been successfully processed as some consumed messages may not have
+    # been yielded yet. To do this, the caller should set
+    # `enable.auto.offset.store` to false and pass processed messages to
+    # {store_offset}. It is also possible, though more complex, to set
+    # 'enable.auto.commit' to false and then pass a manually assembled
+    # TopicPartitionList to {commit}.
+    #
+    # As with `each`, iteration will end when the consumer is closed.
+    # Exception behavior is also the same as with `each`.
+    #
+    # @param max_items [Integer] Maximum size of the yielded array of messages
+    #
+    # @param timeout_ms [Integer] Timeout of this poll
+    #
+    # @raise [RdkafkaError] When polling fails
+    #
+    # @yieldparam message [Batch] An array of received messages
+    #
+    # @return [nil]
+    def each_batch(max_items: 100, timeout_ms: 250, &block)
+      slice = []
+      loop do
+        break if @closing
+        start_time = Time.new.to_f
+        end_time = start_time + timeout_ms / 1000.0
+        max_wait = end_time - Time.now.to_f
+        max_wait_ms = if max_wait <= 0
+                        0   
+                      else
+                        (max_wait * 1000).floor
+                      end 
+        message = nil
+        begin
+          message = poll max_wait_ms
+        rescue Rdkafka::RdkafkaError => error
+          # https://docs.confluent.io/2.0.0/clients/librdkafka/rdkafka_8h.html
+          #   RD_KAFKA_RESP_ERR_REBALANCE_IN_PROGRESS = 27,
+          if error.rdkafka_response == 27
```

Comment:
@mgrosso Theoretically, we could receive any error during the calls to `poll`.
What do you think about yielding the batch followed by raising the received error?

If we do not do this, we might fail the at least once test in "auto everything" scenario.

Reply:
Hi @Adithya-copart , I'm happy to do that but two things about it trouble me:

1. Tell me if I'm missing something here but I don't think we can offer *any* guarantee to the "auto-everything" consumer and it should be discouraged for any use case that needs one.

My reasoning is that the client code processing of the message can take an arbitrary amount of time before raising an uncaught exception, being zapped by the kernel oom handler, or losing power to the node.

At that random point when message processing fails for the "auto everything" consumer the asynchronous offset management may or may not have stored the offset or committed it to the broker.  If it *has*, then there will be an unprocessed message.

However in the opposite case, suppose the processing code successfully handled one or more messages but they haven't been committed yet. When another consumer is assigned this partition they will duplicate the processing.

I see this issue as fundamental to "auto everything" rather than unique to `each_batch`. ie, if we do it here, we should also do it in `each`.  I can see how yielding the batch for "normal exceptions" helps the "auto-everything" consumer get closer to "almost-always-once" so I think it makes sense for rebalance. 

2. My second concern is that passing the outstanding buffered messages in many if not most of the other error possibilities will just guarantee duplicate processing for consumers who manage their own commits or offsets.  Many of those errors names suggest that any subsequent attempt to commit will fail. eg: RD_KAFKA_RESP_ERR_KAFKA_STORAGE_ERROR
https://github.com/edenhill/librdkafka/blob/39796d359898c07ea422849e6d7cd34cd13ec466/src/rdkafka.h#L260
Of course that's not horrible for "at-least-once" in and of itself, but it doesn't seem especially helpful to me either. 

Do those issues make sense?  Do you think it's worth it instead to identify a subset of exceptions worth yielding in both `each` and `each_batch`? that seems daunting to me, and like it might require maintenance over time.

Thanks again for all your careful help with this PR.

Reply:
> My reasoning is that the client code processing of the message can take an arbitrary amount of time before raising an uncaught exception, being zapped by the kernel oom handler, or losing power to the node.
> 
> At that random point when message processing fails for the "auto everything" consumer the asynchronous offset management may or may not have stored the offset or committed it to the broker. If it has, then there will be an unprocessed message.

As long as the library delivers the message to the client, we should be upholding the at least once delivery guarantee.
If the client fails to process it, the responsibility rests with the client but not the library.

I do agree that the actual exception while accumulating the buffer will be obscured if it is followed by another exception in the client code if we yield it. If it happens in client space, clients should be handling it.

For clients not using auto-commit, if a consumer processing a message fails to commit the offset, the client can choose to  undo the processing or re-process the message again on a subsequent delivery. Other client space exceptions can be handled as per the clients choosing.

> However in the opposite case, suppose the processing code successfully handled one or more messages but they haven't been committed yet. When another consumer is assigned this partition they will duplicate the processing.

I think duplicate processing is okay if the at least once guarantee is upheld by `rdkafka`. Consumers that need at-most once can always use manual commits by ensuring the processing of the message/batch and committing the offsets is atomic. This would prevent any duplicate processing.

> Do you think it's worth it instead to identify a subset of exceptions worth yielding in both each and each_batch?

I don't think `rdkafka` needs to change anything in the `each` case. We call `poll` whose result could be either a message or an exception. The client receives it directly and there is no buffer involved.

In the case of `each_batch`, we are introducing a buffer which is not available in the client space until it is yielded. So, if we do not yield this buffer in case of an exception, the client will never receive it but could be marked as processed which is the case currently if we other `Enumerable` methods other than `each`.

Reply:
> Thanks again for all your careful help with this PR.

Thanks for your kind words Matt. Glad I could be of help.
I'm learning quite a bit from you in return as well.

Reply:
Interesting discussion!

I do think the offsets should in principle only be added if the batch was actually completed. Maybe we should disable automatic offset storage by default and set the offsets to the one of the latest message after the `yield`?

Reply:
Hi @thijsc, when proposing to set `enable.auto.offset.store = false` by default, what scope do you mean? How could we keep the functionality intact for existing users of the library?

Reply:
> When proposing to set enable.auto.offset.store = false by default, what scope do you mean?

The default config is set [here](https://github.com/appsignal/rdkafka-ruby/blob/main/lib/rdkafka/config.rb#L57).

> How could we keep the functionality intact for existing users of the library?

I could see us also calling `store_offset` in `each`. Then it would set the offset after a successful yield, which is more robust anyway. The only thing that will then be affected is people using `poll` manually. There are probably not many, if any at all out there. We could add a warning in the release notes and document it.